Refactor campaignion tracking

alex-mo opened the pull request.

alex-mo committed:
campaignion_tracking: Basic structure.

- Drupal module boilerplate + JS boilerplate
- build with parcel

alex-mo committed:
campaignion_tracking: TrackerManager (js).

- simple PubSub structure for tracking events in Campaignion
- provide means for storing information to the sessionStorage
- some tests

alex-mo committed:
campaignion_tracking: fn to parse URL fragment.

alex-mo committed:
campaignion_tracking: issue events for some parts.

- just hardcode some prefixes to get going
- usable: "gtm"

alex-mo committed:
campaignion_tracking_gtm: Basic structure (GTM).

- Drupal module boilerplate + JS boilerplate
- rename main.js --> common.js
- rename some npm scripts according to the 2 entrypoints

alex-mo committed:
campaignion_tracking_gtm: simple listener.

alex-mo committed:
campaignion_tracking: printDebug(), public debug, topics.

alex-mo committed:
campaignion_tracking: refactor into main.js, listener.js.

- rename common.js to main.js again (as main.js is the only entrypoint
  now, use for dev)
- new module for listeners
- read debugging state from sessionStorage
- publish on channel `code`

alex-mo committed:
campaignion_tracking: Basic event handling for GTM.

alex-mo committed:
campaignion_tracking: Add gtm to main.js.

alex-mo committed:
campaignion_tracking: Provide information as Drupal.setting.

- different tracking contexts are defined: node, webform, donation
- fragments in the URL are used to indicate a submission

alex-mo committed:
campaignion_tracking: JS for triggering tracking messages.

alex-mo committed:
campaignion_tracking: Parcel builds for Drupal.

- campaignion_tracking_common.min.js and campaignion_tracking_gtm.min.js
  should be separate files
- campaignion_tracking_gtm.min.js depends on
  campaignion_tracking_common.min.js being already loaded

alex-mo committed:
campaignion_tracking: Update bundled tracking JS.

alex-mo committed:
campaignion_tracking: Implement sending events to GTM.

- use a context store in the tracker.
- calculate revenue, overrideable with eventData
- currencyCode optional in GTM event

alex-mo committed:
rename to campaignion_tracking_common.min.js

torotil commented:
Where can I find the source of the bundled tracking JS? Has it changed?

torotil commented on campaignion_tracking/package/src/fragment.js:
```diff
@@ -0,0 +1,90 @@
+ *
+ * @returns {Object[]} A list of matched items.
+ */
+export function parseLocationHashForPrefixes (prefixes = [], locationHash = '') {
```
Is this actually used in production code?

alex-mo replied:
not yet.

alex-mo replied:
is now.

torotil commented on campaignion_tracking/package/src/gtm.js:
```diff
@@ -1,34 +1,143 @@
+     * The tracker listens for specific parts in the URL fragement and
+     * published messages with event codes to trigger.
+     *
+     * See the listener.js file for this listener.
```
which listener? It’s not clear what “this listener” is referring to in this sentence.

alex-mo replied:
well, there is only one listener in the referenced file, but yes the FragmentListener is the one referenced to. Could be more explicit...

torotil commented on campaignion_tracking/package/src/gtm.js:
```diff
@@ -1,34 +1,143 @@
+     * Subscribe to fragment events.
+     *
+     * The tracker listens for specific parts in the URL fragement and
+     * published messages with event codes to trigger.
```
Why does it need to listen for both? Isn’t the TrackerManager supposed to turn fragments and on-page events into a unified source of pubsub-messages?

alex-mo replied:
No. The TrackerManager currently should only turn fragments into code messages. It does not care what they mean, just adds an prefix. The interested trackers themselves are responsible to listen to them. So they can decide not to be interested in them.

Could be changed of course...

alex-mo replied:
move the code listener to the common module for the Drupal integration and changed the messages chnnel to just `code` (not eg. `code:gtm`) any more...

torotil commented on campaignion_tracking/package/src/main.js:
```diff
@@ -11,13 +12,14 @@ import * as listener from './listener'
 
 // some out-of-the-box prefixes
-export const codePrefixes = []
+export const codePrefixes = ['gtm']
```
Why are there tracker specific fragment prefixes?

alex-mo replied:
The reason was to be easily integratable with the facebook pixel code which is working that way.

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -13,3 +22,193 @@ function campaignion_tracking_preprocess_html(&$vars) {
+/**
+ * Load a payment.
+ *
+ * TODO: is there a little_helper for that?
```
No. It’s hardly the responsibility of little_helpers. You might be interested in [`$submission->payments` provided by webform_paymethod_select](https://github.com/moreonion/webform_paymethod_select/blob/530b79a9d1dfa6710f1bc681229ccdc65221b76a/webform_paymethod_select.module#L392-L436)

You might need [this workaround from campaignion_logcrm](https://github.com/moreonion/campaignion_logcrm/blob/0869ba2089ecc6fac9004ab5ff9078056ba36afe/integrations/webform_paymethod_select.inc#L15-L20) though.

alex-mo replied:
will take a look!

alex-mo replied:
changed it

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -13,3 +22,193 @@ function campaignion_tracking_preprocess_html(&$vars) {
+ *
+ * TODO: Maybe also do this in node_view
+ */
+function campaignion_tracking_page_build(&$page) {
```
Shouldn’t this be made obsolete by the fragment appended to the redirect URL?

alex-mo replied:
gone now.

the fragment triggers a message on the trackermanager and the gtm tracker uses it's sessionStorage state to get the information (along with additional information from the fragment like the sid and paymethod)

torotil commented on campaignion_tracking/js/campaignion_tracking.js:
```diff
@@ -0,0 +1,123 @@
+          variant: String(donation['interval']),
+          quantity: 1
+        };
+        // TODO change name, place
```
I don’t understand what this `TODO` wants to tell me.

alex-mo replied:
ah, yes, this was referring to a prior code version. was an annotation to think again about the domain language and where to put the creation of the `product` object)

torotil commented on campaignion_tracking/js/campaignion_tracking.js:
```diff
@@ -0,0 +1,123 @@
+   * - 'checkoutBegin'
+   * - 'checkoutEnd'
+   */
+  $('.webform-client-form', context).once('campaignion-tracking', function () {
```
The `once()` hinders any event from being dispatched that is not in the `settings` the first time the behaviour is called. I don’t think this is the intended behaviour.

alex-mo replied:
this is called once per page load (when not using webform AJAX) or once per AJAX form submit.
AJAX form submits completely replace the `<form>` markup, thus this handler get's called exactly once per (successful) submit on multistep forms.

The reason for this once is that otherwise the behaviour get's called *multiple* times *per* 1 submit.

Maybe there is a better way. But it does what it is supposed to to (as far as I have tested)...

alex-mo replied:
reworked. `once()` is gone now. introduced `settings['campaignion_tracking']['sent']` to be sure we are not sending the same events twice...

torotil reviewed (changes requested):
That’s a big one. Let’s huddle on Monday.

alex-mo committed:
campaignion_tracking: debugging pushes to (existing) queues.

alex-mo committed:
campaignion_tracking: add queue debug to main.js.

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -0,0 +1,175 @@
+ * Implements hook_preprocess_html().
+ */
+function campaignion_tracking_preprocess_html(&$vars) {
+  $module_path = drupal_get_path('module', 'campaignion_tracking');
```
In general I prefer `#attached` over `drupal_add_*`. This would require moving this into `hook_page_build()` or something similar.

alex-mo replied:
Is there a reason to prefer `#attached`?

torotil replied:
Yes. It makes your function a pure function instead of having side-effects, which means you might be able to actually test it.

alex-mo replied:
If that is the goal, then you cannot use `$element['#attached']['js'][]` but need define a key, I guess. Otherwise it's not pure...

torotil replied:
You should be able to translate this directly into:

```php
$element['#attached'][$path] = $options;
```

See [drupal_process_attached()](https://devdocs.io/drupal~7/includes-common.inc/function/drupal_process_attached/7.x)

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -0,0 +1,175 @@
+  $is_donation = $type && $type->isDonation();
+
+  // General information about the node.
+  drupal_add_js([
```
This should use: `$node->content['#attached']` (or something similar).

alex-mo replied:
changed

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -0,0 +1,175 @@
+  $webform = Webform::fromNode($form['#node']);
+
+  // Information about the current webform.
+  drupal_add_js([
```
Please use `$form['#attached']` here instead of `drupal_add_js()`.

alex-mo replied:
ok

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -0,0 +1,175 @@
+ *
+ * Add webform tracking data to the URL-fragment of the redirect.
+ */
+function campaignion_tracking_add_webform_fragment($form, &$form_state) {
```
This should be done using `hook_webform_redirect_alter()` defined in `little_helpers` — otherwise it won’t work with custom redirects.

alex-mo replied:
changed

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -0,0 +1,175 @@
+ *
+ * Add donation success tracking data to the URL-fragment of the redirect.
+ */
+function campaignion_tracking_add_donation_fragment($form, &$form_state) {
```
Also needs to be done in `hook_webform_redirect_alter()` to work with custom redirects.

alex-mo replied:
changed

torotil commented on campaignion_tracking/js/campaignion_tracking.js:
```diff
@@ -0,0 +1,82 @@
+    var donation = settings.campaignion_tracking.context['donation'] || {};
+    var webform = settings.campaignion_tracking.context['webform'] || {};
+
+    if (node['is_donation']) {
```
Would it be better to dispatch the events during the submit-events on the form? — Maybe something to be discussed broken out into another ticket.

alex-mo replied:
yes, agree, would be better. also yes, maybe best in a new card...

torotil commented on campaignion_tracking/package/src/drupal-common.js:
```diff
@@ -0,0 +1,84 @@
+fragmentListener.setup()
+
+/**
+ * Subscribe to messages of the `code:gtm` tracking channel.
```
The comment seems to be outdated (still referring to `gtm`).

alex-mo replied:
fixed

torotil commented on campaignion_tracking/package/src/gtm.js:
```diff
@@ -0,0 +1,335 @@
+    return gtmData
+  }
+
+  updateContext (context) {
```
How is this context different from the tracker independent context? What additional information does it contain?

alex-mo replied:
Not a all. It's here mainly because there was not need (yet) to have it in the tracker independent place... :)

So it's missing refactoring...

torotil reviewed (changes requested):
Much better already!

* Happy to discuss my questions ;-)
* Still needs a bit of coverage for the PHP changes I guess.

alex-mo committed:
campaignion_tracking: listener for fragments.

alex-mo committed:
campaingion_tracking: prefer #attached over drupal_add_js.

alex-mo committed:
campaignion_tracking: refactor.

- use hook_webform_redirect_alter
- use webform_paymethod_select_webform_submission_load

alex-mo committed:
campaignion_tracking: only set context for full view_mode.

alex-mo committed:
campaignion_tracking: basic tests.
